nx/lib/nx/defn/evaluator.ex

```diff
@@ -42,135 +42,304 @@ defmodule Nx.Defn.Evaluator do
+    {[last_cache | clauses_cache], {all_ids, cache}} =
+      Enum.map_reduce(clause_caches, {%{}, cache}, fn clause_cache, seen_ids_cache ->
+        {clause_cache, seen_ids_cache} =
+          Enum.flat_map_reduce(clause_cache, seen_ids_cache, fn
+            {id, %_{} = tensor}, {seen_ids, cache} ->
+              case seen_ids do
+                # We have already processed this id for the whole cond
+                %{^id => _} ->
+                  {[], {seen_ids, cache}}
+
+                # Process the ID which exists outside of cond
+                %{} ->
+                  cache = composite_compute_cache(tensor, state, cache)
+                  {[], {Map.put(seen_ids, id, true), cache}}
+              end
+
+            {id, counter}, seen_ids_cache ->
+              {[{id, counter}], seen_ids_cache}
+          end)
+
+        {Map.new(clause_cache), seen_ids_cache}
+      end)
+
+    Map.put(cache, [:cond | id], {clauses_cache, last_cache, Map.keys(all_ids)})
+  end
+
+  defp compute_cache(:token, %{data: %Expr{args: [token], id: id}}, state, cache) do
+    hooks = state.hooks
+
+    {hooks, cache} =
+      Enum.map_reduce(token.hooks, cache, fn
+        %{callback: callback, expr: expr, name: name}, cache ->
+          hook_fun = hooks[name] || callback
+
+          cond do
+            hook_fun -> {hook_fun, composite_compute_cache(expr, state, cache)}
+            Tree.has_hooks?(expr, hooks) -> {true, composite_compute_cache(expr, state, cache)}
+            true -> {false, cache}
+          end
+      end)
+
+    Map.put(cache, [:token | id], hooks)
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :elem, args: args}}, state, cache) do
-    [tuple, i] = args
-    {tuple, cache} = composite_eval(tuple, state, cache)
-    {elem(tuple, i), cache}
+  defp compute_cache(_op, tensor, state, cache) do
+    {_, acc} = Tree.apply_args(tensor, cache, &{&1, compute_cache(&1, state, &2)})
+    acc
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :attach_token, args: [token, expr]}}, state, cache) do
-    {_, cache} = eval(token, state, cache)
-    eval(expr, state, cache)
+  ## Evaluation
+
+  defp eval(%Nx.Tensor{data: %Expr{op: :tensor, args: [t]}}, _state, caches) do
+    {t, caches}
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: :metadata, args: [expr, _meta]}}, state, cache) do
-    eval(expr, state, cache)
+  defp eval(%Nx.Tensor{data: %Expr{op: :metadata, args: [expr, _meta]}}, state, caches) do
+    eval(expr, state, caches)
   end
 
-  defp eval(%Nx.Tensor{data: %Expr{op: op, id: id}} = ans, state, cache) do
+  defp eval(%Nx.Tensor{data: %Expr{op: op, id: id}} = ans, state, [cache | caches]) do
     case cache do
-      %{^id => res} ->
-        {res, cache}
+      %{^id => count} when is_integer(count) ->
+        {res, [cache | caches]} = eval_apply(op, ans, state, [cache | caches])
+        state.gc && :erlang.garbage_collect(self())
+        {res, [decrement_cache(cache, id, count, res) | caches]}
```

Comment:
Why don't we garbage collect after decrementing (possibly pruning) the cache?

Also, why not garbage collect in the branch below?

Reply:
The value that is being pruned from the cache is `res`, which we are returning from this function, which means it is still alive. In another word, we purge from the catch the last time a value will be used, but it still has to be used.

Reply:
Ah, makes sense!